storybookjs/storybook
# Addon-docs: Rich prop tables

Issue: #8523 #8484

Prerequisite for new UI: #8828

## What I did

Refactor the PropDef tree structure to accept a string or a summary/detail data structure for the `type` and `defaultValue`.

Improved the rendering of React PropTypes and TS props table:

**Type**
- Large types are replaced by a smart summary and an expandable detail view.
- Display the name of the React component when available instead of "element".
- Object literal, Func, Class, Array formatting has been improved.

**Default Value**
- Large default values are replaced by a smart summary and an expandable detail view.
- The quotes are gone.
- "null" and "undefined" have been replaced by "-".
- Display the name of the function when available instead of "func".
- Display the name of the React component when available instead of "element".
- Object literal, Func, Class, Array formatting has been improved.

## How to test

Run the Jest tests for the following files:

- jsdocParser.test.ts
- generateFuncSignature.test.ts
- acornParser.test.ts

And go to the following path `?path=/docs/docgen-types--page` of `cra-ts-kichen-sink`.

NOTE: Additional tests will be added once the beta is available. They have been removed to make the cut!

- Is this testable with Jest or Chromatic screenshots? Yes
- Does this need a new example in the kitchen sink apps? No
- Does this need an update to the documentation? No

If your answer is yes to any of these, please make sure to include it in your PR.
